### ISSUE TITLE
combineCols() error

### ISSUE THREAD

**Issue by gpertea** [Bioconductor/S4Vectors#102]
I ran into some crashes recently while trying to use `SummarizedExperient::combineCols()`/`combineUniqueCols()` and I traced the failure to this line in S4Vectors:::

https://github.com/Bioconductor/S4Vectors/blob/9ae758483668eca015ef1833704c825fd84dfbec/R/DataFrame-combine.R#L228

The problem seems to be deeper in `DataFrame` subsettting and it is triggered by the presence of a `CharacterList` column in the rowData() `DataFrame` object. 

The following code reproduces this DataFrame subsetting bug:

```R
> DF=DataFrame(id=c("A", "B"), lst=CharacterList(0), row.names = c("A","B"))
> DF
DataFrame with 3 rows and 2 columns
           id             lst
  <character> <CharacterList>
A           A               0
B           B               0
C           C               0

> DF[c("A","B","C"), ,drop=F]
Error: subscript contains NAs
```
The bug can be worked around in the context linked above by converting to data.frame and back, which is obviously not a proper solution:
```R
DataFrame(as.data.frame(DF)[c("A","B","C"), ,drop=F])
DataFrame with 3 rows and 2 columns
            id    lst
   <character> <list>
A            A      0
B            B      0
NA          NA       
```

**Comment by hpages** [comment]
Hi,

The root of the problem is that, unlike ordinary lists, CharacterList objects (and List derivatives in general) don't support subsetting by a subscript with NAs or with out-of-range subscripts at the moment:
```
library(IRanges)

x <- list(a=c("tuv", "wx"), b="yz")
y <- CharacterList(x)

x[2:1]
# $b
# [1] "yz"
#
# $a
# [1] "tuv" "wx" 

y[2:1]
# CharacterList of length 2
# [["b"]] yz
# [["a"]] tuv wx
 
x[c(2:1, NA)]
# $b
# [1] "yz"
# 
# $a
# [1] "tuv" "wx" 
# 
# $<NA>
# NULL

y[c(2:1, NA)]
# Error: subscript contains NAs

x[c(2:1, 3)]
# $b
[1] "yz"
# 
# $a
# [1] "tuv" "wx" 
# 
# $<NA>
# NULL

y[c(2:1, 3)]
# Error: subscript contains out-of-bounds indices
```
I'll work on this but it might take some time because it touches some sensible parts of the **S4Vectors**/**IRanges** code.

In the meantime, one way to work around this is to avoid using CharacterList objects (or other List derivatives) in the DataFrame objects you want to combine:
```
DF <- DataFrame(id=c("A", "B"), X=I(x), row.names=c("A", "B"))

DF[c("A","B","C"), ]
# DataFrame with 3 rows and 2 columns
#               id      X
#      <character> <list>
# A              A tuv,wx
# B              B     yz
# <NA>          NA       

combineCols(DF, DataFrame(Y=11:12, row.names=c("A", "C")))
# DataFrame with 3 rows and 3 columns
#            id      X         Y
#   <character> <list> <integer>
# A           A tuv,wx        11
# B           B     yz        NA
# C          NA               12
```
Obviously not a satisfying situation :disappointed: 

H.

**Comment by gpertea** [comment]
Thank you for all your great work, Hervé, and really appreciate your reply here, I learned from it! There is no problem to "cleanup" the DataFrame before combining, i.e. convert `Compressed*List` columns to plain lists - did not know that the `I()` function can do that as well.  

The `combineCols()/combineUniqueCols()` methods added to SummarizedExperiment are such a godsend that just having to carefully prepare the input `DataFrame`s like this is nothing compared to the pain we had to go through before, in order to merge large RSE objects in memory.
